### PR TITLE
stop suppressing warnings in notebooks

### PR DESCRIPTION
**Pull request by jameslamb**
This PR follows up on a discussion in https://github.com/saturncloud/examples/pull/41#pullrequestreview-531827222. In that PR, @jsignell argued that we should never filter out warnings in these examples, because they're opportunities to improve our code or the upstream libraries we depend on. I totally agree.

This PR proposes:

* remove any calls to filter out warnings
* fixes some warnings in notebooks
* add CI check that will prevent PRs that filter warnings from being merged

The new validation code produces output like this:

```text
------ check results ------

12 errors found checking examples

1. Found use of warnings.simplefilter() in /home/jlamb/repos/saturn/examples/examples/examples-gpu/nyc-taxi-snowflake/rf-rapids.ipynb. Do not filter out warnings in example notebooks. Try to fix them or add text explaining why they can be safely ignored.
2. Found use of warnings.simplefilter() in /home/jlamb/repos/saturn/examples/examples/examples-gpu/nyc-taxi-snowflake/rf-scikit.ipynb. Do not filter out warnings in example notebooks. Try to fix them or add text explaining why they can be safely ignored.
3. Found use of warnings.simplefilter() in /home/jlamb/repos/saturn/examples/examples/examples-gpu/nyc-taxi-snowflake/rf-rapids-dask.ipynb. Do not filter out warnings in example notebooks. Try to fix them or add text explaining why they can be safely ignored.
4. Found use of warnings.simplefilter() in /home/jlamb/repos/saturn/examples/examples/examples-gpu/nyc-taxi/rf-rapids.ipynb. Do not filter out warnings in example notebooks. Try to fix them or add text explaining why they can be safely ignored.
5. Found use of warnings.simplefilter() in /home/jlamb/repos/saturn/examples/examples/examples-gpu/nyc-taxi/rf-scikit.ipynb. Do not filter out warnings in example notebooks. Try to fix them or add text explaining why they can be safely ignored.
6. Found use of warnings.simplefilter() in /home/jlamb/repos/saturn/examples/examples/examples-gpu/nyc-taxi/rf-rapids-dask.ipynb. Do not filter out warnings in example notebooks. Try to fix them or add text explaining why they can be safely ignored.
7. Found use of warnings.simplefilter() in /home/jlamb/repos/saturn/examples/examples/examples-cpu/nyc-taxi-snowflake/hyperparameter-scikit.ipynb. Do not filter out warnings in example notebooks. Try to fix them or add text explaining why they can be safely ignored.
8. Found use of warnings.simplefilter() in /home/jlamb/repos/saturn/examples/examples/examples-cpu/nyc-taxi-snowflake/hyperparameter-dask.ipynb. Do not filter out warnings in example notebooks. Try to fix them or add text explaining why they can be safely ignored.
9. Found use of warnings.simplefilter() in /home/jlamb/repos/saturn/examples/examples/examples-cpu/nyc-taxi/hyperparameter-scikit.ipynb. Do not filter out warnings in example notebooks. Try to fix them or add text explaining why they can be safely ignored.
10. Found use of warnings.simplefilter() in /home/jlamb/repos/saturn/examples/examples/examples-cpu/nyc-taxi/dashboard.ipynb. Do not filter out warnings in example notebooks. Try to fix them or add text explaining why they can be safely ignored.
11. Found use of warnings.simplefilter() in /home/jlamb/repos/saturn/examples/examples/examples-cpu/nyc-taxi/hyperparameter-dask.ipynb. Do not filter out warnings in example notebooks. Try to fix them or add text explaining why they can be safely ignored.
12. Found use of warnings.simplefilter() in /home/jlamb/repos/saturn/examples/examples/examples-cpu/nyc-taxi/data-aggregation.ipynb. Do not filter out warnings in example notebooks. Try to fix them or add text explaining why they can be safely ignored.
```

## "A value is being set on a slice from a DataFrame"

<details><summary>full traceback</summary>

```text
/srv/conda/envs/saturn/lib/python3.7/site-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/srv/conda/envs/saturn/lib/python3.7/site-packages/ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/srv/conda/envs/saturn/lib/python3.7/site-packages/ipykernel_launcher.py:9: FutureWarning: Series.dt.weekofyear and Series.dt.week have been deprecated.  Please use Series.dt.isocalendar().week instead.
  if __name__ == '__main__':
/srv/conda/envs/saturn/lib/python3.7/site-packages/ipykernel_launcher.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  if __name__ == '__main__':
/srv/conda/envs/saturn/lib/python3.7/site-packages/ipykernel_launcher.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  # Remove the CWD from sys.path while we load stuff.
/srv/conda/envs/saturn/lib/python3.7/site-packages/ipykernel_launcher.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  # This is added back by InteractiveShellApp.init_path()
/srv/conda/envs/saturn/lib/python3.7/site-packages/ipykernel_launcher.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  if sys.path[0] == '':
```

</details>

This is caused by this pattern in our code

```python
df = df[df.fare_amount > 0]
```

Because the dataframe is passed in by reference, each of these other subsetting things is like the "chained assignment" referred to in https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy. The easiest solution is just to do a `df.copy()` inside this function. Since we're assigning the result of `prep_df()` to a new data frame anyway, this doesn't change the peak memory usage for our notebooks, and it resolves the warning.

## "Please use Series.dt.isocalendar()"

<details><summary>full traceback</summary>

```text
/srv/conda/envs/saturn/lib/python3.7/site-packages/ipykernel_launcher.py:9: FutureWarning: Series.dt.weekofyear and Series.dt.week have been deprecated.  Please use Series.dt.isocalendar().week instead.
  if __name__ == '__main__':
```

</details>

This comes from feature-engineering code in `prep_df()`. Changing it like the warning suggests is enough to make this warning go away.

### CODE REVIEW

File: .ci/validate-examples.py [hunk 1]
```diff
@@ -111,6 +112,36 @@ def image_exists_on_dockerhub(image_name: str, image_tag: str) -> bool:
     return res.status_code == 200
 
 
+def _lint_python_cell(file_name: str, code_lines: List[str]) -> List[str]:
+    """
+    Given the content of a Python code cell, check if for problems we
```
Comment (jsignell) [inline .ci/validate-examples.py]:
```suggestion
    Given the content of a Python code cell, check it for problems we
```

File: examples/examples-cpu/nyc-taxi/hyperparameter-dask.ipynb [hunk 2]
```diff
@@ -141,7 +139,7 @@
     "    df['tip_fraction'] = df.tip_amount / df.fare_amount\n",
     "    \n",
     "    df['pickup_weekday'] = df.tpep_pickup_datetime.dt.weekday\n",
```
Comment (jsignell) [inline examples/examples-cpu/nyc-taxi/hyperparameter-dask.ipynb:141]: I think this one needs to be changed as well
```suggestion
    "    df['pickup_weekday'] = df.tpep_pickup_datetime.dt.isocalendar().day\n",
```

**Review by jsignell — CHANGES_REQUESTED** [review]
Thank you for taking this on! I just think you missed the `dt.weekday` warnings.

File: examples/examples-cpu/nyc-taxi/hyperparameter-dask.ipynb (hunk 2 above)
Comment (jameslamb) [reply examples/examples-cpu/nyc-taxi/hyperparameter-dask.ipynb:141]: oh, `weekday` hasn't been deprecated. Only `.weekofyear` and `.week` are mentioned in the warning.

> FutureWarning: Series.dt.weekofyear and Series.dt.week have been deprecated.  Please use Series.dt.isocalendar().week instead.

When I run the `nyc-taxi/xgboost.ipynb` notebook, for example, no warnings are generated even though it uses `.dt.weekday`.

Comment (jsignell) [reply examples/examples-cpu/nyc-taxi/hyperparameter-dask.ipynb:141]: You are 100% right. Good lord I am not doing a good job reading that warning at all :facepalm:

**Review by jsignell — APPROVED** [review]

File: examples/examples-cpu/nyc-taxi/hyperparameter-dask.ipynb (hunk 2 above)
Comment (jameslamb) [reply examples/examples-cpu/nyc-taxi/hyperparameter-dask.ipynb:141]: haha it's ok.

This is why I prefer `df["column_name"]` to `df.column_name`  in `pandas` code, but I don't think starting that argument is a good use of our time right now :joy:

**Pull request merged**